app/presenters/application_presenter.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+class ApplicationPresenter < SimpleDelegator
+  include ActionView::Helpers::FormTagHelper
+  include ActionView::Helpers::UrlHelper
+  include Rails.application.routes.url_helpers
+
+  attr_reader :action, :controller
+
+  # Master list of fields that could be displayed on any page.
+  ALL_FIELDS = %i[].freeze
+
+  # Not displayed on the show page
+  HIDDEN_FIELDS = %i[].freeze
+
+  # Displayed on the edit page but not editable
+  READ_ONLY_FIELDS = %i[].freeze
+
+  # Only editable when creating a new object
+  READ_ONLY_AFTER_PERSISTED_FIELDS = %i[].freeze
+
+  # Default character size of editable fields
+  FIELD_SIZE = 60
+
+  def self.attribute_scope
+    @attribute_scope ||= "activerecord.attributes." + name.sub(%r{Presenter$}, "").underscore
+  end
+
+  def self.value_scope
+    @value_scope ||= name.sub(%r{Presenter$}, "").underscore + ".values"
+  end
+
+  def self.field_label(field)
+    I18n.t(field, scope: attribute_scope)
+  end
+
+  def initialize(obj, controller: nil, action: :show)
+    super obj
+    @controller = controller
+    @action = action
+  end
+
+  # By default form cancel button goes to show page if persisted,
+  # index page otherwise.
+  def cancel_path
+    if persisted?
+      method = (self.class.name.sub(%r{Presenter$}, "").underscore + "_path").to_sym
+      send method, self
+    else
+      method = (self.class.name.sub(%r{Presenter$}, "").pluralize.underscore + "_path").to_sym
+      send method
+    end
+  end
+
+  def editable?(field, form: nil)
+    return false if form.nil?
+    return false if self.class::READ_ONLY_FIELDS.include?(field)
+    return false if persisted? && self.class::READ_ONLY_AFTER_PERSISTED_FIELDS.include?(field)
+
+    true
+  end
+
+  def field_label(field, form: nil)
+    return "" if self.class::HIDDEN_FIELDS.include?(field) && action == :show
+
+    if editable?(field, form: form)
+      form.label field
+    else
+      self.class.field_label(field)
+    end.to_s.html_safe
+  end
+
+  def field_value(field, form: nil)
+    return "" if self.class::HIDDEN_FIELDS.include?(field) && action == :show
+
+    if editable?(field, form: form)
+      edit_field_value(field, form: form)
+    else
+      show_field_value field
+    end.to_s.html_safe
+  end
+
+  private
+
+  def show_field_value(field)
+    method = ("show_" + field.to_s).to_sym
+    if respond_to?(method, true)
+      send(method) || ""
+    else
+      localize_value field
+    end
+  end
+
+  def localize_value(field)
+    value = try(field) || return
```

Comment:
This will call the method `field` in the object if it exists; we're not sure what the use case there is.

Reply:
I think I can explain the reason I used `try(f)` instead of `send(f)`: for derived or aggregate fields that are not model methods (for example the composite `auth_rep` in `HTRegistration` that is composed of the three model name/email/date fields) when the presenter's `show_X` method hasn't been written yet. (The `show_X` bleeds off the `localize_value` logic.) So `send(:auth_rep)` would throw an exception where `try(:auth_rep)` would return `nil`, and `localize_value` would just return `nil` at that point. So the value of `try` is to get something running in a development scenario where the presenter may not be fleshed out. I'll leave it up to @aelkiss whether to comment and keep,  or replace with `send`.

Reply:
TL;DR
If my presenter lists (in e.g., `ALL_FIELDS`) a field `:blah` and we're rendering it.
1. Call `self.show_blah` if it's implemented (implementation is responsible for getting the value).
2. Otherwise call `self.try :blah` or `self.send :blah` (the latter potentially throwing an exception) to get the value.
3. Check for a localized version of the value with `t(self.class.value_scope + ".#{field}.#{value}"`
4. HTML escape the raw value if all else fails (which is perfectly legit).

Reply:
Removed in forthcoming commit.